### PR TITLE
user guide: developing

### PR DESCRIPTION
**Pull request by oliver-sanders**
Merge after #6 

Add some notes for documentation writers.

Also includes a few technical changes:

* All admonitions are now linkable (including practical and solution boxes)
* Admonitions can now be used in CLI documentation (e.g.: `NOTE: Note here` is interpretted as `.. note:: Note here`).
* Fixed the link-symbol for titled code-blocks in the rtd theme.
* Any doctests in sphinx documentation are run in the rose test-battery.

### CODE REVIEW

File: sphinx/ext/auto-cli-doc.py [hunk 1]
```diff
@@ -342,6 +413,24 @@ def write_rose_command_reference(write, command_name):
...
+    """Generate help text for a rose command.
+
+    Args:
+        command_name (str): The name of the rose command to be documented (i.e.
+            rose / rosie.
+
+    """
+    # Obtain help text.
+    cmd = [command_name, 'doc']
+    stdout = Popen(cmd, stdout=PIPE).communicate()[0]
```
Comment (matthewrmshin) [inline sphinx/ext/auto-cli-doc.py]: Should die on zero return code?

**Review by matthewrmshin — COMMENTED** [review]
One minor comment. Looks OK otherwise.

File: sphinx/ext/auto-cli-doc.py (hunk 1 above)
Comment (oliver-sanders) [reply sphinx/ext/auto-cli-doc.py]: Done.

**Review by matthewrmshin — APPROVED** [review]

File: sphinx/developing/restructuredtext.rst [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+   Sub-Sub Heading
+   ^^^^^^^^^^^^^^^
+
+   Totally Irrelevant Heading
+   """"""""""""""""""""""""""
+
+For consistency:
+
+* Top level headings allways appear at the top of the document (only one per
```
Comment (dvalters) [inline sphinx/developing/restructuredtext.rst]: 'allways' -> 'always'

File: sphinx/developing/restructuredtext.rst [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+Headings are text lines which are underlined. The underline character can be
+any of:
+
+.. code-block:: none
+
+   ! " # $ % & ' ( ) * + , - . / : ; < = > ? @ [ \ ] ^ _ ` { | } ~
+
+Different characters are used to define different document levels e.g. section,
+sub-section, etc. reStructuredText doesn't actually care which character you
+use for each role it infers precedence from the order of appearance.
```
Comment (dvalters) [inline sphinx/developing/restructuredtext.rst]: some sort of punctuation here after role: '...use for each role it infers precedence...'

File: sphinx/developing/auto-cli-api.rst [hunk 4]
```diff
@@ -0,0 +1,167 @@
+Auto CLI API
+============
+
+
+The rose command line API is auto-documented by the extension ``auto-cli-api``.
```
Comment (dvalters) [inline sphinx/developing/auto-cli-api.rst:5]: is rose capitalised? (not sure what the convention is...)

File: sphinx/developing/auto-cli-api.rst [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+This translates the rose CLI help into reStructuredText. To keep the CLI docs
+clean they are written in a plain format which has some RreStructuredText
```
Comment (dvalters) [inline sphinx/developing/auto-cli-api.rst]: -> 'reStructuredText'

File: sphinx/developing/autodoc.rst [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+Writing Docstrings
+------------------
+
+.. _Google Docstrings: https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments
+.. _Numpy Docstrings: https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt
+.. _Napoleon Google Example: http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html
+.. _Napoleon Numpy Example: http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html
+
+This documentation uses ``sphinx.ext.napoleon`` to pre-process docstrings
+allowing them to be written in plain text as opposed to ReStructuredText.
```
Comment (dvalters) [inline sphinx/developing/autodoc.rst]: 'reStructructuredText' - I think it just needs to be made consistent throughout the docs (elsewhere it is 'RST' and ReStructuredText')

**Review by dvalters — COMMENTED** [review]
Looks good - very minor comments/suggestions (typos/consistency etc)

**Comment by dvalters** [comment]
Couldn't find it here but in the `rose make-docs --help` page it says "cashing" when I think it should be "caching" in that context.

File: sphinx/developing/auto-cli-api.rst (hunk 4 above)
Comment (oliver-sanders) [reply sphinx/developing/auto-cli-api.rst:5]: So rose is supposedly a contrived acronym but we now treat it as a name, which we don't capitalise? @matthewrmshin ...

File: sphinx/developing/autodoc.rst (hunk 6 above)
Comment (dvalters) [reply sphinx/developing/autodoc.rst]: there's a typo in my typo correction...

**Review by dvalters — APPROVED** [review]

**Pull request merged**